### DISCUSSION TITLE
Trouble for Compiling

### DISCUSSION THREAD

**Discussion by anawwaaf** [grblHAL/RP2040#89] [Q&A]
I have a 3 axis CNC machine for milling and I want to use a Raspberry Pico RP2040 as a controller.
I've done a git clone
```
git clone --recurse-submodules https://github.com/grblHAL/RP2040.git
```

and I have edited `my_machine.h` and added `my_machine_map.h`

edited section in `my_machine.h` :
```
#define BOARD_MY_MACHINE                              //in line 33
#define SPINDLE0_ENABLE         SPINDLE_PWM0          //in line 48
#define EEPROM_ENABLE          32                     //in line 78
```

this is the `my_machine_map.h` based on generic_map.h :

```
#if TRINAMIC_ENABLE
#error Trinamic plugin not supported!
#endif

#if N_ABC_MOTORS
#error "Axis configuration is not supported!"
#endif

// Define step pulse output pins.
#define STEP_PORT               GPIO_PIO  // N_AXIS pin PIO SM
#define STEP_PINS_BASE          2         // N_AXIS number of consecutive pins are used by PIO

// Define step direction output pins.
#define DIRECTION_PORT          GPIO_OUTPUT
#define X_DIRECTION_PIN         5
#define Y_DIRECTION_PIN         6
#define Z_DIRECTION_PIN         7
#define DIRECTION_OUTMODE       GPIO_SHIFT5

// Define stepper driver enable/disable output pin.
#define ENABLE_PORT             GPIO_OUTPUT
#define STEPPERS_ENABLE_PIN     8

// Define homing/hard limit switch input pins.
#define X_LIMIT_PIN             9
#define Y_LIMIT_PIN             10
#define Z_LIMIT_PIN             11
#define LIMIT_INMODE            GPIO_MAP

// Define driver spindle pins

#if DRIVER_SPINDLE_PWM_ENABLE
#define SPINDLE_PWM_PORT        GPIO_OUTPUT
#define SPINDLE_PWM_PIN         15
#else
#define AUXOUTPUT3_PORT         GPIO_OUTPUT
#define AUXOUTPUT3_PIN          15
#endif

#if DRIVER_SPINDLE_DIR_ENABLE
#define SPINDLE_PORT            GPIO_OUTPUT
#define SPINDLE_DIRECTION_PIN   14
#else
#define AUXOUTPUT4_PORT         GPIO_OUTPUT
#define AUXOUTPUT4_PIN          14
#endif

#if DRIVER_SPINDLE_ENABLE
#ifndef SPINDLE_PORT
#define SPINDLE_PORT            GPIO_OUTPUT
#endif
#define SPINDLE_ENABLE_PIN      13
#else
#define AUXOUTPUT5_PORT         GPIO_OUTPUT
#define AUXOUTPUT5_PIN          13   
#endif

// Define flood and mist coolant enable output pins.
#define COOLANT_PORT            GPIO_OUTPUT
#define COOLANT_FLOOD_PIN       16
#define COOLANT_MIST_PIN        17

// Define auxillary I/O
#define AUXINPUT0_PIN           22
#define AUXINPUT1_PIN           21
#define AUXINPUT2_PIN           23
#define AUXINPUT3_PIN           28 // Probe

#define AUXOUTPUT0_PORT         GPIO_OUTPUT
#define AUXOUTPUT0_PIN          12
#define AUXOUTPUT1_PORT         GPIO_OUTPUT
#define AUXOUTPUT1_PIN          26
//#define AUXOUTPUT2_PORT         GPIO_OUTPUT
//#define AUXOUTPUT2_PIN          27
#define NEOPIXELS_PIN 27
//#define NEOPIXELS_NUM 5
// Define user-control controls (cycle start, reset, feed hold) input pins.

#define RESET_PIN               18
#define FEED_HOLD_PIN           19
#define CYCLE_START_PIN         22

#if PROBE_ENABLE
#define PROBE_PIN               AUXINPUT3_PIN
#endif

#if I2C_ENABLE
#define I2C_PORT            0
#define I2C_SDA             20
#define I2C_SCL             21
#endif

#if SAFETY_DOOR_ENABLE
#define SAFETY_DOOR_PIN         AUXINPUT1_PIN
#endif

#if MOTOR_FAULT_ENABLE
#define MOTOR_FAULT_PIN         AUXINPUT0_PIN
#endif
```

to set the compatibility level, 
```
#if !defined COMPATIBILITY_LEVEL || defined __DOXYGEN__
#define COMPATIBILITY_LEVEL 0   
#endif
```

do I just need to change this `#define COMPATIBILITY_LEVEL 0 ` in `config.h` ? 

do I need to add my board to the `driver.json` file?
`driver.json` :
```
{
  "caps":
  {
    "informal": 1,
    "axes": 6,
    "usb_cdc": 1,
    "uart": 1,
    "spindle_pwm": 1,
    "spindle_dir": 1,
    "serial_ports": 2,
    "eeprom": 1,
    "i2c_strobe": 1,
    "ganged_axes": 1,
    "auto_square": 1,
    "digital_in": 8,
    "digital_out": 8,
    "probe": 1,
    "safety_door": 1,
    "estop": 1,
    "mpg_mode": 1,
    "sdcard": 1,
    "wifi": 1,
    "wiznet": 1,
    "littlefs": 1,
    "bluetooth": 1,
    "plugins": [
      { "id": "motors" },
      { "id": "spindle" },
      { "id": "modbus" },
      { "id": "eeprom" },
      { "id": "sdcard" },
      { "id": "webui" },
      { "id": "networking" },
      { "id": "keypad" },
      { "id": "macros" },
      { "id": "laser_cluster" },
      { "id": "laser_coolant" },
      { "id": "bluetooth" },
      { "id": "fans" },
      { "id": "embroidery" }
    ]
  },
  "docker_instance": "RP2040",
  "symbols": {
    "OVERRIDE_MY_MACHINE": 1
  },
  "boards": [
    {
      "name": "PicoCNC",
      "symbol": "BOARD_PICO_CNC",
      "URL": "https://github.com/phil-barrett/PicoCNC",
      "MAP": "pico_cnc_map.h",
      "caps": {
        "axes": 5,
        "digital_in": 4,
        "digital_out": 8,
        "i2c_strobe": 1,
        "eeprom": 1,
        "fram": 1,
        "i2c": 1,
        "sdcard": 1,
        "wifi": 1,
        "wiznet": 1,
        "bluetooth": 1
      }
    },
    {
      "name": "PicoBOB",
      "symbol": "BOARD_PICOBOB",
      "URL": "https://github.com/Expatria-Technologies/PicoBOB",
      "MAP": "picobob_map.h",
      "caps": {
        "axes": 5,
        "safety_door": 0,
        "serial_ports": 0
      }
    },
    {
      "name": "PicoBOB_G540",
      "symbol": "BOARD_PICOBOB_G540",
      "URL": "https://github.com/Expatria-Technologies/PicoBOB",
      "MAP": "picobob_g540_map.h",
      "caps": {
        "axes": 4,
        "safety_door": 0,
        "serial_ports": 0
      }
    },
    {
      "name": "PicoBOB_DLX",
      "symbol": "BOARD_PICOBOB_DLX",
      "URL": "https://github.com/Expatria-Technologies/PicoBOB-DLX",
      "MAP": "picobob_dlx_map.h",
      "caps": {
        "axes": 5,
        "digital_in": 2,
        "digital_out": 3,
        "i2c_strobe": 1,
        "safety_door": 0,
        "serial_ports": 0,
        "wiznet": 1
      }
    },
    {
      "name": "PicoBOB_DLX_G540",
      "symbol": "BOARD_PICOBOB_DLX",
      "URL": "https://github.com/Expatria-Technologies/PicoBOB-DLX",
      "MAP": "picobob_dlx_g540_map.h",
      "caps": {
        "axes": 4,
        "digital_in": 2,
        "digital_out": 3,
        "i2c_strobe": 1,
        "safety_door": 0,
        "serial_ports": 0,
        "wiznet": 1
      }
    },    
    {
      "name": "BTT SKR Pico 1.0",
      "symbol": "BOARD_BTT_SKR_PICO_10",
      "URL": "https://github.com/bigtreetech/SKR-Pico",
      "MAP": "btt_skr_pico_10_map.h",
      "caps": {
        "axes": 4,
        "trinamic": 2209,
        "trinamic_uart_addr": 1
      }
    },
	{
      "name": "Generic",
      "symbol": "",
      "MAP": "generic_map.h",
      "caps": {
        "axes": 3,
        "digital_in": 1,
        "digital_out": 3,
        "i2c_strobe": 1,
        "wifi": 1,
        "bluetooth": 1
      }
    },
    {
      "name": "Generic 4 axis",
      "symbol": "BOARD_GENERIC_4AXIS",
      "MAP": "generic_map_4axis.h",
      "caps": {
        "axes": 4,
        "i2c_strobe": 1,
        "wifi": 1,
        "bluetooth": 1
      }
    }
  ]
}
```

**Comment by terjeio** [comment]
> do I just need to change this #define COMPATIBILITY_LEVEL 0 in config.h ?

Only if you are using a sender that does not like the [grblHAL extensions](https://github.com/grblHAL/core/wiki/Compatibility-level).

> do I need to add my board to the driver.json file?

No, this file is only used by the Web Builder.

**Reply by anawwaaf** [reply]
thanks
I have successfully compiled grblHAL with `kit: Pico ARM GCC`. and drag-drop the `grblHAL.uf2` file onto the board. But, when I use UGS Platform or io Sender, it can't connect
UGS Platform:
```
*** Connecting to jserialcomm://COM10:115200
*** Fetching device status
>>> ?
*** Fetching device status (2 of 10)...
>>> ?
*** Fetching device status (3 of 10)...
>>> ?
*** Fetching device status (4 of 10)...
>>> ?
*** Fetching device status (5 of 10)...
>>> ?
*** Fetching device status (6 of 10)...
>>> ?
*** Fetching device status (7 of 10)...
>>> ?
*** Fetching device status (8 of 10)...
>>> ?
*** Fetching device status (9 of 10)...
>>> ?
*** Fetching device status (10 of 10)...
>>> ?
**** Connection closed ****
```

ioSender:
![Sender_connection_config](https://github.com/grblHAL/RP2040/assets/98207229/b66833ef-06b8-4866-8ea0-0771966c1006)
![Sender_response](https://github.com/grblHAL/RP2040/assets/98207229/731921af-8231-404e-b0e9-01833717083d)

I also tried compiling with other board settings, and only generic board settings were able to connect

I've also tried the firmware from the web builder but it only works for generic boards. If I use another board option, the result is the same as not being able to connect

**Reply by terjeio** [reply]
> I've also tried the firmware from the web builder but it only works for generic boards.

I have just built for PicoCNC and PicoBOB and both images works for me.
You have a bad USB cable or USB driver?
If you connect with a terminal program such as PuTTY do you get any response to commands such as `?` and `$i` ?

**Reply by anawwaaf** [reply]
Yeah, I replaced the cable and it works but I had a problem when changing grbl settings

for EEPROM, I used a DS1307 RTC module, I connected this RTC module to the RP2040, and this caused grbl not to work if I made changes to settings such as invert limit pin and so on

**Reply by terjeio** [reply]
The DS1307 has 56 bytes of memory, default for grblHAL is 2K - absolute minimum is 1K which is not enough if any plugins that has settings is added to the build.

**Reply by anawwaaf** [reply]
![hw111](https://github.com/grblHAL/RP2040/assets/98207229/fa24b9c0-319a-4a45-9cc5-a005df5d8b14)
Sorry, I use this module, HW-111 I2C module, I want to use AT24C32 in this module for EEPROM

**Reply by terjeio** [reply]
Is the AT24C32 configured for the [correct address](https://github.com/grblHAL/Plugin_EEPROM/blob/fd157b303e2c37b1be1a329177ef34fecc4f49fa/eeprom_24AAxxx.c#L33)?

**Reply by anawwaaf** [reply]
AT24C32 address is 0x50
eeprom_24AAxxx.c :
```
#define EEPROM_I2C_ADDRESS (0xA0 >> 1)
```
should I change this to 0x50 ?

**Reply by terjeio** [reply]
> should I change this to 0x50 ?

No, it is 0x50 already (0x50 = 0xA0 shifted right one bit).

I assume you uncommented and set [#define EEPROM_ENABLE](https://github.com/grblHAL/RP2040/blob/339e6450c965c3b0e2c6d41e30fe74b7c5ed75ac/my_machine.h#L78) to 32 already. If so it should work provided the wiring is correct - and it looks like there are pullup resistors on the SCL and SDA pins.
You may check the voltage on SLC and SDA, both should be ~3.3V - if not the bus is stuck for some reason.

**Reply by anawwaaf** [reply]
I've checked the voltage on SCL and SDA,
SCL = ~3.316 V
SDA = ~3.125 V

**Reply by terjeio** [reply]
Ok then the bus is not stuck. FYI I have checked my Pico CNC board which has a 64Kbit EEPROM and that is working fine.
The 32Kbit code is the same as the 64Kbit code so I assume the code is ok. So you have a wiring issue or a bad EEPROM?

**Reply by anawwaaf** [reply]
This module (EEPROM) works fine when I use it on Arduino Uno. 

For wiring, I use a jumper, directly from the module to RP2040
| HW-111 | RP2040 |
| -------- | ------- |
| VCC  | 3V3 |
| GND | GND |
| SDA | GP20 |
| SCL | GP21 |

**Reply by terjeio** [reply]
`$pins` output is?

**Reply by anawwaaf** [reply]
![RP2040_HW-111](https://github.com/grblHAL/RP2040/assets/98207229/aa0a92c4-5306-4d87-ac77-098b905b7ccf)
Wiring for HW-111 module

> `$pins` output is?

What does it mean?

**Reply by terjeio** [reply]
If you send the `$pins` command from the MDI (or from a terminal connection) the pin assignments are listed to the console. Use it to verify that they are correct.
FYI I am currently away, will be back next week. Until then I cannot check the code with a 32Kbit EEPROM - and I am not sure I have any so I may have to order a few.

**Reply by anawwaaf** [reply]
output for `$pins` :
```
>>> $pins
[PIN:19,Feed hold]
[PIN:22,Cycle start]
[PIN:18,Emergency stop]
[PIN:28,Probe]
[PIN:9,X limit min]
[PIN:10,Y limit min]
[PIN:11,Z limit min]
[PIN:22,Aux in 0,P0]
[PIN:21,Aux in 1,P1]
[PIN:23,Aux in 2,P2]
[PIN:PIO2,X step]
[PIN:PIO3,Y step]
[PIN:PIO4,Z step]
[PIN:5,X dir]
[PIN:6,Y dir]
[PIN:7,Z dir]
[PIN:8,Steppers enable]
[PIN:13,Spindle on]
[PIN:14,Spindle direction]
[PIN:15,Spindle PWM]
[PIN:17,Mist]
[PIN:16,Flood]
[PIN:12,Aux out 0,P0]
[PIN:26,Aux out 1,P1]
[PIN:27,LED adressable]
[PIN:21,I2C SCK]
[PIN:20,I2C SDA]
[PIN:1,RX,Primary UART]
[PIN:0,TX,Primary UART]
```

**Reply by terjeio** [reply]
You have a pin conflict: `[PIN:21,Aux in 1,P1]
`

**Reply by anawwaaf** [reply]
I've tried changing the I2C pin to
SDA 16
SCL 17

and this works, I can adjust the grbl settings.

**Reply by terjeio** [reply]
20/21 should work too if you remove the Aux in 1 conflict.

**Reply by anawwaaf** [reply]
> You have a pin conflict: `[PIN:21,Aux in 1,P1] `

yeah, thanks for reminding me,

I tried to correct it, and I can use I2C on pins 20 and 21 too

**Reply by anawwaaf** [reply]
what are the differents between
```
#define AUXINPUT1_PIN 28

#if PROBE_ENABLE
#define PROBE_PIN AUXINPUT1_PIN
#endif
```
And
```
#if PROBE_ENABLE
#define PROBE_PIN 28
#endif
```

**Reply by anawwaaf** [reply]
and if I try
```
#define AUXINPUT0_PIN 21

#if I2C_ENABLE
#define I2C_PORT 0
#define I2C_SDA 20
#define I2C_SCL AUXINPUT0_PIN
#endif
```

and the result is that GRBL doesn't work, if I change the GRBL settings from MDI, GRBL doesn't respond

but if I define `#define I2C_SCL 21` and remove the `#define AUXINPUT0_PIN 21`, GRBL works fine

```
#if I2C_ENABLE
#define I2C_PORT 0
#define I2C_SDA 20
#define I2C_SCL 21
#endif
```

**Reply by terjeio** [reply]
The difference is that the former leaves the pin as an aux input if the probe input is not enabled.

Allocating pins from the aux pool is not made available for pins connected to peripherals like I2C or SPI since they are typically routed to on-board devices.
